Remove multi-file DartPad usages from cheatsheet (#5392)
dart-lang/site-www: src/codelabs/dart-cheatsheet.md

String stringify(int x, int y) {
  TODO('Return a formatted string here');
}

// Tests your solution (Don't edit!): 
void main() {
  assert(stringify(2, 3) == '2 3',
      "Your stringify method returned '${stringify(2, 3)}' instead of '2 3'");
  print('Success!');
}

> {$ end main.dart $}

Reviewer: Can you get rid of these `begin` and `end main.dart` tags across the file? Support for them will also be going away and they're not necessary with just one Dart file anyway.

Author: Oh, sure! Didn't know

  }

  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }

  // ignore_for_file: unnecessary_type_check
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for collection literals example</b>
  </summary>

  This exercise is fairly straightforward.
  Just add a list, set, or map literal after each equals sign.

>   See the codelab text for the correct syntax to use.

Reviewer: I know you didn't write this, but while we're here, can we remove this text that makes it sound easy? It might not be easy for someone.

Also the "see the codelab text..." portion doesn't make sense anymore with the answer right below.

I can't really think of any super valuable hint here, if you can't either, I'd be fine just dropping it. Perhaps we could mention something like "Just remember to specify the types when they can't be inferred." or something like that for the empty ones?

I'll come back to this comment if I think of anything else. Thanks!!

Author: I put:

    Add a list, set, or map literal after each equals sign.
    Remember to specify the types for the empty declarations,
    since they can't be inferred.

I think it's ok just as a description of the solution, not necessarily a hint

    if (product != 30) {
      errs.add('The product property returned $product \n instead of the expected value (30).'); 
    } 
  } catch (e) {
    print('Tried to use MyClass.product, but encountered an exception: \n ${e.runtimeType}.');
    return;
  }

  try {
    obj.incrementValue1();
    
    if (obj.value1 != 3) {
      errs.add('After calling incrementValue, value1 was ${obj.value1} \n instead of the expected value (3).'); 
    } 
  } catch (e) {
    print('Tried to use MyClass.incrementValue1, but encountered an exception: \n ${e.runtimeType}.');
    return;
  }

  try {
    final joined = obj.joinWithCommas(['one', 'two', 'three']);
    
    if (joined != 'one,two,three') {
      errs.add('Tried calling joinWithCommas([\'one\', \'two\', \'three\']) \n and received $joined instead of the expected value (\'one,two,three\').'); 
    } 
  } catch (e) {
    print('Tried to use MyClass.joinWithCommas, but encountered an exception: \n ${e.runtimeType}.');
    return;
  }

  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for arrow syntax example</b>
  </summary>

>   For the product, you can just multiply the three values together.

Reviewer: Maybe we can be a bit more specific? As long as we drop "just" I'll be a happy camper.

```suggestion
  For the product, you can use `*` to multiply the three values together.
```

  For `incrementValue1`, you can use the increment operator (`++`).

>   For `joinWithCommas`, try using the `join` method found in the `List` class.

Reviewer: Since this is in the solution rather than a hint now, "try" doesn't make sense anymore.

```suggestion
  For `joinWithCommas`, use the `join` method found in the `List` class.
```

    } 
  } on UnimplementedError {
    _result(false, ['Tried to call joinWithCommas but failed. \n Did you implement the method?']);
    return;
  } catch (e) {
    _result(false, ['Tried calling joinWithCommas(1, 2 ,3), \n but encountered an exception: ${e.runtimeType}.']);
    return;
  }

  try {
    final value = joinWithCommas(1, 2, 3, 4, 5);
    
    if (value != '1,2,3,4,5') {
      errs.add('Tried calling joinWithCommas(1, 2, 3, 4, 5) \n and got $value instead of the expected (\'1,2,3,4,5\').'); 
    } 
  } on UnimplementedError {
    _result(false, ['Tried to call joinWithCommas but failed. \n Did you implement the method?']);
    return;
  } catch (e) {
    _result(false, ['Tried calling stringify(1, 2, 3, 4 ,5), \n but encountered an exception: ${e.runtimeType}.']);
    return;
  }

  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for positional parameters example</b>
  </summary>
  The `b`, `c`, `d`, and `e` parameters are null if they aren't provided by the

>   caller. The important thing, then, is to check whether those arguments are null

Reviewer: ```suggestion
  caller. The important thing, then, is to check whether those arguments are `null`
```

    tryFunction(() {}, logger);
  
    if (logger.lastType != null) {
      errs.add('Untrustworthy didn\'t throw an Exception, \n but one was logged anyway: ${logger.lastType}.');
    }
    
    if (logger.lastMessage != '') {
      errs.add('Untrustworthy didn\'t throw an Exception with no message, \n but a message was logged anyway: \'${logger.lastMessage}\'.');
    }
    
    if (!logger.done) {
      errs.add('Untrustworthy didn\'t throw an Exception, \n but doneLogging() wasn\'t called afterward.');
    }
  } catch (e) {
    _result(false, ['Untrustworthy didn\'t throw an exception, \n but an exception of type ${e.runtimeType} was unhandled by tryFunction anyway.']);
  }
  
  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for exceptions example</b>
  </summary>
  This exercise looks tricky, but it's really one big `try` statement.

>   Just call `untrustworthy` inside the `try`, and

Reviewer: ```suggestion
  Call `untrustworthy` inside the `try`, and
```

  final String aString;
  final double aDouble;
  

>   // TODO: Create a constructor here.

Reviewer: Should this be "the constructor" since we're instructed them to create a specific one?

```suggestion
  // TODO: Create the constructor here.
```

}

// Tests your solution (Don't edit!):
void main() {
  final errs = <String>[];
  
  try {
    final obj = MyClass(1, 'two', 3);
    
    if (obj.anInt != 1) {
      errs.add('Called MyClass(1, \'two\', 3) and got an object with anInt of ${obj.anInt} \n instead of the expected value (1).');
    }

    if (obj.anInt != 1) {
      errs.add('Called MyClass(1, \'two\', 3) and got an object with aString of \'${obj.aString}\' \n instead of the expected value (\'two\').');
    }

    if (obj.anInt != 1) {
      errs.add('Called MyClass(1, \'two\', 3) and got an object with aDouble of ${obj.aDouble} \n instead of the expected value (3).');
    }
  } catch (e) {
    _result(false, ['Called MyClass(1, \'two\', 3) and got an exception \n of type ${e.runtimeType}.']);
  }
  
  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for `this` example</b>
  </summary>
  This exercise has a one-line solution.

>   Just declare the constructor with

Reviewer: ```suggestion
  Declare the constructor with
```

  `this.anInt`, `this.aString`, and `this.aDouble`
  as its parameters in that order.

  ```dart    

>     MyClass(this.anInt, this.aString, this.aDouble);

Reviewer: No need to indent here since it's alone in the snippet.

```suggestion
MyClass(this.anInt, this.aString, this.aDouble);
```

  }
  
  if (!caughtException) {
    errs.add('Called FirstTwoLetters(\'\') and didn\'t get an exception \n from the failed assertion.');
  }
  
  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for initializer lists example</b>
  </summary>
  Two assignments need to happen:

>   `letterOne` should be `word[0]`, and `letterTwo` should be `word[1]`.

Reviewer: Do you think it's worth being more specific here?

```suggestion
  `letterOne` should be assigned to `word[0]`, and
  `letterTwo` should be assigned to `word[1]`.
```

Author: Agree with specifying the action, but wouldn't it be more correct to say it the other way around? Like, I'd read that code as "`word[0]` is assigned to `letterOne`", so I guess the instruction would be:

`letterOne` should be assigned `word[0]`

I put this in my edit but lmk if that's less accurate

  
  Color(this.red, this.green, this.blue);

  // TODO: Create a named constructor called "Color.black" here:

}

// Tests your solution (Don't edit!):
void main() {
  final errs = <String>[];

  try {
    final result = Color.black();
    
    if (result.red != 0) {
      errs.add('Called Color.black() and got a Color with red equal to \n ${result.red} instead of the expected value (0).');
    }

    if (result.green != 0) {
      errs.add('Called Color.black() and got a Color with green equal to \n ${result.green} instead of the expected value (0).');
    }

    if (result.blue != 0) {
  errs.add('Called Color.black() and got a Color with blue equal to \n ${result.blue} instead of the expected value (0).');
    }
  } catch (e) {

>     _result(false, ['Called Color.black() and got an exception of type \n ${e.runtimeType}.']);

Reviewer: Looks like this step still need to be migrated away from `_result`.

    return;
  }

  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for named constructors example</b>
  </summary>

>   The declaration for your constructor should be `Color.black() {}`.
>   Inside the braces, set `red`, `green`, and `blue` to zero.

Reviewer: We must have not updated this hint when it was changed to use an initializer list for null safety.

How does this sound?

```suggestion
  The declaration for your constructor should begin with `Color.black(): `.
  In the initializer list (after the colon), set `red`, `green`, and `blue` to `0`.
```

Author: Nice catch!

  } on Error {
    _throwed = true;
  } catch (e) {

>     _result(false, ['Called IntegerSingle.fromList([]) and got an exception of \n type ${e.runtimeType}.']);

Reviewer: This sample also needs to be updated away from `_result`.

    final obj = IntegerHolder.fromList([1, 2, 3]);
    
    if (obj is! IntegerTriple) {
      errs.add('Called IntegerHolder.fromList([1, 2, 3]) and got an object of type \n ${obj.runtimeType} instead of IntegerTriple.');
    } else {
      if (obj.a != 1) {
        errs.add('Called IntegerHolder.fromList([1, 2, 3]) and got an IntegerTriple \n with an \'a\' value of ${obj.a} instead of the expected (1).');
      }
      
      if (obj.b != 2) {
        errs.add('Called IntegerHolder.fromList([1, 2, 3]) and got an IntegerTriple \n with an \'a\' value of ${obj.b} instead of the expected (2).');
      }

      if (obj.c != 3) {
        errs.add('Called IntegerHolder.fromList([1, 2, 3]) and got an IntegerTriple \n with an \'a\' value of ${obj.b} instead of the expected (2).');
      }
    }
  } catch (e) {
    _result(false, ['Called IntegerHolder.fromList([1, 2, 3]) and got an exception \n of type ${e.runtimeType}.']);
    return;
  }

  if (errs.isEmpty) {
    print('Success!');
  } else {
    errs.forEach(print);
  }
}
{$ end main.dart $}
```

<details markdown="1">
 <summary>
   <b>Solution for factory constructors example</b>
  </summary>
  Inside the factory constructor,

>   check the length of the list and create an

Reviewer: ```suggestion
  check the length of the list, then create and return an
```

  List<String> ingredients;
  int calories;
  double milligramsOfSodium;

  // TODO: Create a const constructor here"

}

// Tests your solution (Don't edit!):
void main() {
  final errs = <String>[];

  try {
    const obj = Recipe(['1 egg', 'Pat of butter', 'Pinch salt'], 120, 200);
    
    if (obj.ingredients.length != 3) {
      errs.add('Called Recipe([\'1 egg\', \'Pat of butter\', \'Pinch salt\'], 120, 200) \n and got an object with ingredient list of length ${obj.ingredients.length} rather than the expected length (3).');
    }
    
    if (obj.calories != 120) {
      errs.add('Called Recipe([\'1 egg\', \'Pat of butter\', \'Pinch salt\'], 120, 200) \n and got an object with a calorie value of ${obj.calories} rather than the expected value (120).');
    }
    
    if (obj.milligramsOfSodium != 200) {
      errs.add('Called Recipe([\'1 egg\', \'Pat of butter\', \'Pinch salt\'], 120, 200) \n and got an object with a milligramsOfSodium value of ${obj.milligramsOfSodium} rather than the expected value (200).');
    }
  } catch (e) {

>     _result(false, ['Tried calling Recipe([\'1 egg\', \'Pat of butter\', \'Pinch salt\'], 120, 200) \n and received a null.']);

Reviewer: This sample also needs to be updated away from `_result`.
